docs/src/examples/plasticity.jl

```diff
@@ -0,0 +1,410 @@
+    # Define material parameters
+    E = 200.0e9 # [Pa]
+    H = E/20   # [Pa]
+    ν = 0.3     # [-]
+    σ₀ = 200e6  # [Pa]
+    material = J2Plasticity(E, ν, σ₀, H)
+
+    L = 10.0 # beam length [m]
+    w = 1.0  # beam width [m]
+    h = 1.0  # beam height[m]
+    n_timesteps = 10
+    u_max = zeros(n_timesteps)
+    traction_magnitude = 1.e7 * range(0.5, 1.0, length=n_timesteps)
+
+    # Create geometry, dofs and boundary conditions
+    n = 2
+    nels = (10n, n, 2n) # number of elements in each spatial direction
+    P1 = Vec((0.0, 0.0, 0.0))  # start point for geometry
+    P2 = Vec((L, w, h))        # end point for geometry
+    grid = generate_grid(Tetrahedron, nels, P1, P2)
+    interpolation = Lagrange{3, RefTetrahedron, 1}() # Linear tet with 3 unknowns/node
+
+    dh = create_dofhandler(grid, interpolation) # JuaFEM helper function
+    dbcs = create_bc(dh, grid) # create Dirichlet boundary-conditions
+
+    cellvalues, facevalues = create_values(interpolation)
+
+    # Pre-allocate solution vectors, etc.
+    n_dofs = ndofs(dh)  # total number of dofs
+    u  = zeros(n_dofs)  # solution vector
+    Δu = zeros(n_dofs)  # displacement correction
+    r = zeros(n_dofs)   # residual
+    K = create_sparsity_pattern(dh); # tangent stiffness matrix
+
+    # Create material states. One array for each cell, where each element is an array of material-
+    # states - one for each integration point
+    nqp = getnquadpoints(cellvalues)
+    states = [[MaterialState() for _ in 1:nqp] for _ in 1:getncells(grid)]
+
+   # states = [MaterialState() for _ in 1:nqp for _ in 1:getncells(grid)]
+   # temp_states = [MaterialState() for _ in 1:nqp for _ in 1:getncells(grid)]
+    #
+    ## Newton-Raphson loop
+    NEWTON_TOL = 1 # 1 N
+    print("\n Starting Netwon iterations:\n")
+
+    for timestep in 1:n_timesteps
+        t = timestep # actual time (used for evaluating d-bndc)
+        traction = Vec((0.0, 0.0, traction_magnitude[timestep]))
+        newton_itr = -1
+        print("\n Time step @time = $timestep:\n")
+        update!(dbcs, t) # evaluates the D-bndc at time t
+        apply!(u, dbcs)  # set the prescribed values in the solution vector
+
+        while true; newton_itr += 1
+
+            if newton_itr > 8
+                error("Reached maximum Newton iterations, aborting")
+                break
+            end
+            K, r = doassemble(cellvalues, facevalues, K, grid, dh, material, u,
+                             states, traction);
+            norm_r = norm(r[JuAFEM.free_dofs(dbcs)])
+
+            print("Iteration: $newton_itr \tresidual: $(@sprintf("%.8f", norm_r))\n")
+            if norm_r < NEWTON_TOL
+                break
+            end
+
+            apply_zero!(K, r, dbcs)
+            Δu = Symmetric(K) \ r
+            u -= Δu
+        end
+
+        # Update all the material states after we have reached equilibrium
+        for cell_states in states
+            foreach(update_state!, cell_states)
```

Comment:
Maybe it is possible to make use of a `previous_states` variable which could be used for the equilibrium iterations. Then we could get rid of  `update_state!` and the `temp_` variables in `MaterialState`, but we would need to copy the state-vector each iteration...

Reply:
Yes, the alternative would be to not store `temp_` stuff in the material state but instead have two arrays with states. The advantage is that the material state can then be made immutable.

Reply:
But the state is modified in `compute_stress_tangent`, so it still has to be mutable?
Or should `compute_stress_tangent` return the `σ`, `D`` and a `new_state`?

Reply:
Just `new_state` would be sufficient, and you extract `σ` and `D` from the new state.

Reply:
Yes, in that case you return `state, D`. I don't think `D` is needed to be stored.

Reply:
So, would this necessarily lead to a speedup, or is this more a question of design?
Modifying a mutable struct vs. creating new immutable structs inside the material routine.

Reply:
It would probably be a bit higher performance since in julia, immutable types can be stored inline in arrays (no pointer indirection).

I think the current way is fine though, we can always refine it later if we want to.